lingua_franca/configuration.py

```diff
@@ -0,0 +1,122 @@
+    get_default_loc
+from lingua_franca.internal import resolve_resource_file
+
+default_global_values = \
+    {
+        'load_langs_on_demand': False
+    }
+
+
+class LangConfig(dict):
+    def __init__(self, lang_code):
+        super().__init__()
+        if lang_code not in get_supported_locs():
+            # DO NOT catch UnsupportedLanguageError!
+            # If this fails, we want to crash. This can *only* result from
+            # someone trying to override sanity checks upstairs. There are no
+            # circumstances under which this should fail and allow the program
+            # to continue.
+            lang_code = get_full_lang_code(lang_code)
+
+        resource_file = resolve_resource_file(f'text/{lang_code}/config.json')
+        try:
+            with open(resource_file, 'r', encoding='utf-8') as i_file:
+                default_values = json.load(i_file)
+            for k in default_values:
+                self[k] = default_values[k]
+        except (FileNotFoundError, TypeError):
+            pass
+
+
+class Config(dict):
+    def __init__(self):
+        super().__init__()
+        self['global'] = dict(default_global_values)
+        for lang in get_active_langs():
+            self.load_lang(lang)
+
+    def load_lang(self, lang):
+        if all((lang not in get_supported_locs(),
+                lang in get_supported_langs())):
+            if lang not in self.keys():
+                self[lang] = {}
+                self[lang]['universal'] = LangConfig(lang)
+
+            full_loc = get_full_lang_code(lang)
+        else:
+            full_loc = lang
+            lang = get_primary_lang_code(lang)
+
+        self[lang][full_loc] = LangConfig(full_loc)
+
+    def _find_setting(self, setting=None, lang=''):
+        if setting is None:
+            raise ValueError("lingua_franca.config requires "
+                             "a setting parameter!")
+
+        setting_available_in = []
+        possible_locs = []
+
+        while True:
+            if setting in self['global']:
+                setting_available_in.append('global')
+            if lang == 'global':
+                break
+
+            lang = lang or get_default_loc()
+
+            if lang in get_supported_langs():
+                possible_locs.append((lang, 'universal'))
+                possible_locs.append((lang, get_full_lang_code(lang)))
+
+            if lang in get_supported_locs():
+                primary_lang_code = get_primary_lang_code(lang)
+                possible_locs.append((primary_lang_code, 'universal'))
+                possible_locs.append(
+                    (primary_lang_code, get_default_loc(primary_lang_code)))
+                possible_locs.append((primary_lang_code, lang))
+
+            for place in possible_locs:
+                if setting in self[place[0]][place[1]]:
+                    setting_available_in.append(place)
+
+            break
+        try:
+            return setting_available_in[-1]
```

Comment:
why do we take the last one? I haven't reasoned this through completely but I assume the order would roughly be:
```
[
global, 
default_lang.universal, 
default_lang.specific, 
other_lang.universal, 
other_lang.specific,
default_locale.universal,
default_locale.specific,
other_locale.universal,
other_locale.specific,
place # which I'm unclear what this is
]
```

Is that right? 

If so it seems like we would want the default lang setting over an other locale setting. But I could be getting the whole thing mixed up.

Given that I need to reason about it we should document what is intended in the docstring and write some tests for it.

Reply:
This is definitely one of the things that will need the most discussion, but you're about right re: priority. The list is just in the opposite order of what you're thinking =P

'place' might not be a great variable name in this context, but it's referring to the possible indices of the desired setting, nothing to do with LF's job. Could be in *this* dict or *this* dict or *that* dict...

`possible_places` becomes a list of dicts to check for the config value. `setting_available_in` is the result, and it's ordered from global to local.

If the setting is found at all, `setting_available_in[-1]` will be, in the following order:

1. The locale passed in the `lang` param, or the default locale for the `lang` param, or the current default locale when no `lang` param is passed
2. Language-wide settings, which start out identical to their "default" locale. For instance, `en`'s default locale (from `get_full_lang_code()` is `en-us`, so English's default language-wide settings are loaded from `en-us`'s config.json
3. "Global," LF-wide settings

If the value isn't present in any of those places, it's unavailable in that language.

Reply:
Perusing stales, rephrased. At runtime, LF config is a dict of dicts. It's got settings for the whole module and settings for each locale. Each language defers to a *particular* locale as its "default" locale. This is configurable. Gez's default English locale will be en-AU, Chance's will be en-US, Aditya might change it up on the fly!

So. We want to `get(setting)`.

1. Given a list of `possible_places` (dicts) where the requested setting *might* be stored, look for it
2. Let the list of dicts in which you *found* the value be `setting_available_in`
3. `setting_available_in` should, thanks to the algorithm that built it, be ordered as follows:
    `[global_value, fallback_lang_value, fallback_loc_value, specified_lang_value, specified_loc_value]` (where "fallback" is the default language, if different from the one specified)
4. Return the last element of `setting_available_in`, representing the "most local" value that conforms with the input query

This way, a "more local" config dict will always override its less-local counterpart, assuming the requested key is present in the more local dict.

Reply:
For the sake of populating each field, let's say I am running with

* Default language English, default locale en-US
* Current user language and locale en/en-AU
* Spanish loaded, default locale es-ES
* Alternate locale es-US also loaded

If I ask for an es-US setting that's fully localized, I expect `setting_available_in` to offer values in the order:

`[en-US, en-AU, es-ES, es-US]`

representing, per the previous comment,

`[fallback_lang_value, fallback_loc_value, specified_lang_value, specified_loc_value]`

but, if any of the config dicts are missing the setting I asked for, they'll be absent from this list and LF will move on with its life, returning instead the next value "up" the chain.

At the end, this bit would return `[en-US, en-AU, es-ES, es-US][-1]`, getting `es-US` and, per the initial function call, causing the getter to return the config value from `es-US`.